python/cudf/cudf/core/reshape.py

```diff
@@ -38,36 +40,66 @@ def _align_objs(objs, how="outer"):
 
         index = objs[0].index
         name = index.name
-        if how == "inner" or isinstance(index, cudf.MultiIndex):
-            for obj in objs[1:]:
-                index = (
-                    cudf.DataFrame(index=obj.index)
-                    .join(cudf.DataFrame(index=index), how=how)
-                    .index
-                )
-                index.name = name
-            return [obj.reindex(index) for obj in objs], False
 
-        else:
-            all_index_objs = [obj.index for obj in objs]
-            appended_index = all_index_objs[0].append(all_index_objs[1:])
-            df = cudf.DataFrame(
-                {
-                    "idxs": appended_index,
-                    "order": cudf.core.column.arange(
-                        start=0, stop=len(appended_index)
-                    ),
-                }
+        if how == "inner":
+            final_index = _get_combined_index(
+                [obj.index for obj in objs], intersect=True, sort=sort
             )
-            df = df.drop_duplicates(subset=["idxs"]).sort_values(
-                by=["order"], ascending=True
+        else:
+            final_index = _get_combined_index(
+                [obj.index for obj in objs], intersect=False, sort=sort
             )
-            final_index = df["idxs"]
-            final_index.name = name
 
-            return [obj.reindex(final_index) for obj in objs], False
+        final_index.name = name
+        return [
+            obj.reindex(final_index)
+            if not final_index.equals(obj.index)
+            else obj
+            for obj in objs
+        ]
     else:
-        return objs, True
+        # This check uses `sort is not False` rather than just `sort=True`
+        # to differentiate between a user-provided `False` value and the
+        # default `None`. This is necessary for pandas compatibility, even
+        # though `True` and `False` are the only valid options from the user.
+        if sort:
+            if not first.index.is_monotonic_increasing:
+                final_index = first.index.sort_values()
+                return [obj.reindex(final_index) for obj in objs]
+        return objs
+
+
+def _get_combined_index(indexes, intersect: bool = False, sort=None):
+    if len(indexes) == 0:
+        index = cudf.Index([])
+    elif len(indexes) == 1:
+        index = indexes[0]
+    elif intersect:
+        sort = True
+        index = indexes[0]
+        for other in indexes[1:]:
+            # Don't sort for every intersection,
+            # let the sorting happen in the end.
+            index = index.intersection(other, sort=False)
+    else:
+        index = indexes[0]
+        if sort is False:
+            sort = False
+        elif sort is None:
+            sort = False if isinstance(index, cudf.StringIndex) else True
+        else:
+            sort = True
```

Comment:
The True/False cases don't need to do anything, right?
```suggestion
        if sort is None:
            sort = False if isinstance(index, cudf.StringIndex) else True
```